### PR TITLE
CONTRIBUTING.md: add a section about including patches

### PR DESCRIPTION
**Pull request by drupol**
Inspired from https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations/missing-patch-comment.md

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Janik-Haag** [comment]
There should be a part about making the effort of upstreaming patches if it makes sense since we should avoid carrying patches if possible.

**Comment by drupol** [comment]
> There should be a part about making the effort of upstreaming patches if it makes sense since we should avoid carrying patches if possible.

Added in the first paragraph.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -61,6 +61,49 @@ Pull requests should not be squash merged in order to keep complete commit messa
...
+While we do accept custom patches, we strongly encourage submitting patches
+upstream whenever possible. This practice means submitting the patches against
+the original software's repository, which has the potential to benefit the
+broader user community.
+
+For out-of-tree patches fetched using `fetchpatch`, avoid linking to GitHub pull
+requests, as these are mutable and can change with each new commit or
+force-push. Instead, use direct links to specific commit patches such as
+https://github.com/owner/repo/commit/sha.patch. This ensures that the reference
+to your patch remains constant and accurate.
```
Comment (nim65s) [inline CONTRIBUTING.md:75]: I agree this ensure the reference remains constant, but we loose the context of the PR. A compromise is
https://github.com/owner/repo/pull/number/commits/sha.patch, which gets redirected to the version you propose

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -61,6 +61,49 @@ Pull requests should not be squash merged in order to keep complete commit messa
...
+file that includes the commit message and the associated metadata along with the
+changes themselves. Remember, a well-structured patch is as critical as the code
+changes it contains.
+
+As part of our ongoing efforts to ensure the clarity, maintainability, and
+quality of our source code, we've implemented a policy that requires all
+contributors to provide explanatory comments with every patch submission.
+
+This means that each time you add a patch to the source code, it is mandatory to
+accompany it with a comment that clearly outlines the purpose of the patch.
```
Comment (hacker1024) [inline CONTRIBUTING.md:88]: Should end with a comma instead of a full stop?

**Review by hacker1024 — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md (hunk 1 above)
Comment (drupol) [reply CONTRIBUTING.md]: Good suggestion. Fixed.

File: CONTRIBUTING.md (hunk 2 above)
Comment (drupol) [reply CONTRIBUTING.md]: Updated, let me know if this is better.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -61,6 +61,51 @@ Pull requests should not be squash merged in order to keep complete commit messa
...
+changes themselves. Remember, a well-structured patch is as critical as the code
+changes it contains.
+
+As part of our ongoing efforts to ensure the clarity, maintainability, and
+quality of our source code, we've implemented a policy that requires all
+contributors to provide explanatory comments with every patch submission.
+
+This means that each time you add a patch to the source code, it is mandatory to
+accompany it with a comment that clearly outlines its purpose, regardless of its
+filename.
```
Comment (Ma27) [inline CONTRIBUTING.md:89]: Also regardless of the description inside the patch itself?
When appropriate, I usually do `git format-patch` for patches that are only relevant for nixpkgs as well and then I add a descriptive commit message there as I do it with all other commits %)

**Comment by drupol** [comment]
@infinisil Feel free to let me know what to do with this PR now that it is done. Do you think this is something that you might re-use in your clean up PR?

**Comment by pbsds** [comment]
If patches are to become the one and true sanctioned approach, then I would strongly suggest creating/recommending tooling to aid maintainers in rebasing and maintaining such patches.

Maintaining patches in nixpkgs is not a fun task, especially for large and active repos. It usually requires you to check out the project repo, apply the patch, then rebase it onto master.

Package repositories like FreeBSD Ports offers [a better patching story](https://docs.freebsd.org/en/books/porters-handbook/slow-porting/#slow-patch):
It quickly allows you to check out a writeable version of the package source, where you can muck about and make changes you want, as long as you remember to `cp file-to-change{,.orig}` first. In nixpkgs, simply building `mypackage.src` doesn't help since the contents of the resulting `result` is read-only, and there is no version control information there to aid the rebase.

The result is that many maintainers favor substitueInPlace over patches, simply because they are far easier to maintain.

**Comment by drupol** [comment]
Hi,

I created this PR just to discuss how things should be done in case of a patch inclusion.

I hear your concerns and I feel that your message is more about whether we use a patch in favor of something else. This is not the point of this here... And I don't have enough experience with Nix to have a point of view yet.

The idea is to formalize how things need to be done when a patch needs to be added in nixpkgs, nothing else, nothing more.

**Comment by infinisil** [comment]
> @infinisil Feel free to let me know what to do with this PR now that it is done. Do you think this is something that you might re-use in your clean up PR?

In PM's with @drupol I replied this:

I think the contents of this PR should definitely be merged into https://nixos.org/manual/nixpkgs/stable/#sec-patches. Of course ideally https://github.com/NixOS/nixpkgs/pull/245243 would be done and merged already, in which case that section would live in `pkgs/README.md`, but that's not yet the case, and it probably takes another week [or more] to get that merged.

If you don't mind waiting a bit this would make my job with that PR easier :P

**Comment by Ma27** [comment]
When I take a look at the PRs referenced in here, e.g. #246838, then I really fail to see how a comment `Fix path to ps2pdf binary` is bringing any additional value when the path to the patch one line below exactly says the same, i.e. `fix-ps2pdf-path.patch`.

In fact, this is a good example of where a patch-file is pretty well-named and the comment above only exists for the sake of a guideline IMHO.

What I want to say is, I'm not sure anymore if this is really good to be a _hard_ rule. If the naming is sufficient and there's nothing else important to say (e.g. "remove this patch on the next major version upgrade") then I don't see any value of adding such a comment. This is also related to my previous point: when writing a good commit-message, you'll have a useful summary as the name of the patch and all the remaining details are in the file itself.

**Comment by Janik-Haag** [comment]
I think work on this can continue now since #245243 got merged

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -61,6 +61,51 @@ Pull requests should not be squash merged in order to keep complete commit messa
...
+
+While we do accept custom patches, we strongly encourage submitting patches
+upstream whenever possible. This practice means submitting the patches against
+the original software's repository, which has the potential to benefit the
+broader user community.
+
+For out-of-tree patches fetched using `fetchpatch`, avoid linking to GitHub pull
+requests, as these are mutable and can change with each new commit or
+force-push. Instead, use direct links to specific commit patches such as
+`https://github.com/owner/repo/pull/number/commits/sha.patch`. This ensures that
```
Comment (Janik-Haag) [inline CONTRIBUTING.md:74]: this link pattern also breaks when some one force pushes to the pr so the commit hash changes or if a commit gets dropped from a pr/moved in to another one.

Comment (pbsds) [reply CONTRIBUTING.md:74]: For multi-commit PRs, the `/compare/base...commit.patch` endpoint is optimal, used in #258431

**Comment by fabian-thomas** [comment]
I want to express concern here that while the `compare` endpoint is definitely the way to go for reproducibility it introduces an issue.

Github patches are inherently unstable as reported here https://github.com/orgs/community/discussions/12531. The commit hashes may change in length when the repository grows, but also just randomly as I explain in more detail [here](https://github.com/orgs/community/discussions/12531#discussioncomment-10133721).

For `commit` endpoint patches this can be fixed by appending `?full_index=1` to the URL. This makes the commit hashes full size. Therefore I would suggest to add that to the guide.

Unfortunately, this doesn't work for the `compare` endpoint as I outline [here](https://github.com/orgs/community/discussions/12531#discussioncomment-10133721). For small projects this doesn't seem to cause issues, but for large projects like nixpkgs it's easy to get changes. I hope Github reacts to the report and fixes this completely by defaulting to `full_index=1`, which seems to be a sane default to me. This would also mean that we can drop adding `?full_index=1` to every url from this guide.

**Comment by hacker1024** [comment]
> I want to express concern here that while the `compare` endpoint is definitely the way to go for reproducibility it introduces an issue.
> 
> Github patches are inherently unstable as reported here https://github.com/orgs/community/discussions/12531. The commit hashes may change in length when the repository grows, but also just randomly as I explain in more detail [here](https://github.com/orgs/community/discussions/12531#discussioncomment-10133721).
> 
> For `commit` endpoint patches this can be fixed by appending `?full_index=1` to the URL. This makes the commit hashes full size. Therefore I would suggest to add that to the guide.
> 
> Unfortunately, this doesn't work for the `compare` endpoint as I outline [here](https://github.com/orgs/community/discussions/12531#discussioncomment-10133721). For small projects this doesn't seem to cause issues, but for large projects like nixpkgs it's easy to get changes. I hope Github reacts to the report and fixes this completely by defaulting to `full_index=1`, which seems to be a sane default to me. This would also mean that we can drop adding `?full_index=1` to every url from this guide.

Isn't fetchpatch [supposed to fix this](https://github.com/NixOS/nixpkgs/blob/633ae2570ec145a008d7703886e78faa24b9c1bd/pkgs/build-support/fetchpatch/default.nix#L1)?

**Comment by fabian-thomas** [comment]
> Isn't fetchpatch [supposed to fix this](https://github.com/NixOS/nixpkgs/blob/633ae2570ec145a008d7703886e78faa24b9c1bd/pkgs/build-support/fetchpatch/default.nix#L1)?

Yep, you are right. Thanks for clarifying. Don't mind my comment then.

**Comment by pbsds** [comment]
> Isn't fetchpatch [supposed to fix this](https://github.com/NixOS/nixpkgs/blob/633ae2570ec145a008d7703886e78faa24b9c1bd/pkgs/build-support/fetchpatch/default.nix#L1)?

* https://github.com/NixOS/nixpkgs/issues/257446
* https://github.com/NixOS/nixpkgs/issues/266556

and patchutils has not received updates in four years

**Pull request closed without merging**